### PR TITLE
Test url previewing

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

**Comment by erikjohnston** [comment]
Oops, need to install the libraries on jenkins

File: tests/51media/20urlpreview.pl [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+            method   => "GET",
+            full_uri => "/_matrix/media/r0/preview_url",
+            params   => {
+               url          => $test_server_info->client_location . "/test.html",
+               access_token => $user->access_token,
+            },
+         ),
+      )->SyTest::pass_on_done( "Preview returned successfully" )
+      ->then( sub {
+         my ( $_, $_, $preview_body ) = @_;
```
Comment (leonerd) [inline tests/51media/20urlpreview.pl]: You probably don't mean `$_` here - assign to `undef` to unpack holes.

File: tests/51media/20urlpreview.pl [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+         assert_eq( $preview_body->{"og:url"}, $OGRAPH_URL );
+         assert_eq( $preview_body->{"matrix:image:size"}, 2239 );
+         assert_eq( $preview_body->{"og:image:height"}, 129 );
+         assert_eq( $preview_body->{"og:image:width"}, 279 );
+
+         $preview_body->{"og:image"} =~ m/^mxc:\/\// or die "Expected mxc url for og:image";
+
+         Future->done( 1 );
+      })->SyTest::pass_on_done( "Preview API returned expected values" )
+
```
Comment (leonerd) [inline tests/51media/20urlpreview.pl]: Extraneous blank line

File: tests/51media/20urlpreview.pl [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            Future->done( $request );
+         })->SyTest::pass_on_done( "URL was fetched" ),
+
+         await_http_request( "/test.png", sub {
+            return 1;
+         })->then( sub {
+            my ( $request ) = @_;
+
+            my $pngdata = read_binary( "$DIR/test.png" );
```
Comment (leonerd) [inline tests/51media/20urlpreview.pl]: I don't see this image added to the PR - did you remember to `git add` it?

Comment (erikjohnston) [inline tests/51media/20urlpreview.pl]: There was already one there :)

File: tests/51media/20urlpreview.pl (hunk 1 above)
Comment (erikjohnston) [inline tests/51media/20urlpreview.pl]: Aha! I couldn't for the life of me remember what you were meant to do there

**Comment by erikjohnston** [comment]
@leonerd PTAL

File: tests/51media/20urlpreview.pl [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            method   => "GET",
+            full_uri => "/_matrix/media/r0/preview_url",
+            params   => {
+               url          => $test_server_info->client_location . "/test.html",
+               access_token => $user->access_token,
+            },
+         ),
+      )->SyTest::pass_on_done( "Preview returned successfully" )
+      ->then( sub {
+         my ( undef, undef, $preview_body ) = @_;
```
Comment (leonerd) [inline tests/51media/20urlpreview.pl]: Hmm. Actually the more I look at it; it seems like you're having the first two parts of the three-way `needs_all` return a request object only to then ignore them here. You could just

    Future->done()

to have those first two return an empty list.

File: tests/51media/20urlpreview.pl [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+            my ( $request ) = @_;
+
+            my $response = HTTP::Response->new( 200 );
+            $response->add_content( $EXAMPLE_OPENGRAPH_HTML );
+            $response->content_type( "text/html" );
+            $response->content_length( length $response->content );
+
+            $request->respond( $response );
+
+            Future->done( );
```
Comment (leonerd) [inline tests/51media/20urlpreview.pl]: As a special-case, empty parens don't have whitespace; `()`

File: tests/51media/20urlpreview.pl [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+            my $pngdata = read_binary( "$DIR/test.png" );
+
+            my $response = HTTP::Response->new( 200 );
+            $response->add_content( $pngdata );
+            $response->content_type( "image/png" );
+            $response->content_length( length $response->content );
+
+            $request->respond( $response );
+
+            Future->done( );
```
Comment (leonerd) [inline tests/51media/20urlpreview.pl]: ditto

**Comment by leonerd** [comment]
LGTM aside from whitespace

**Comment by erikjohnston** [comment]
Ta

**Pull request merged**